### PR TITLE
Corrects for singularity distortion.

### PR DESCRIPTION
**Pull request by TemporalOroboros**
## About the PR

Corrects transformations between the map and the screen to account for singularity distortion visuals.
~~Currently broken, for some reason it only works without errors if the singulo is centered.~~
Fixed: The singularities position was being vertically mirrored to match up with fragment space. I just needed to undo that to get it to work.
Requires https://github.com/space-wizards/RobustToolbox/pull/4188
**Media**

Will add a video showcasing it working in the morning.

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase
https://youtu.be/Mr7Th9wG3nk

**Changelog**

:cl:
- fix: Clicking on the screen now accounts for the distortion effect of nearby singularities.

### CODE REVIEW

**Comment by TemporalOroboros** [comment]
I'll keep working on the visualizer when it isn't 3am.

File: Content.Client/Singularity/SingularityOverlay.cs [hunk 1]
```diff
@@ -86,6 +88,39 @@ protected override void Draw(in OverlayDrawArgs args)
...
+                
+                if(deformation > 0.8)
+                    deformation = MathF.Pow(deformation, 0.3f);
+
+                Vector2 displacement = delta * deformation;
+                finalCoords -= displacement;
+            }
+
+            args.ScreenPosition = finalCoords;
+        }
```
Comment (moonheart08) [inline Content.Client/Singularity/SingularityOverlay.cs:124]: Don't hardcode this here, I believe you could compute a transform matrix for this value, so instead have an event requesting interested parties provide a transform for screen space and handle it wherever client singulo is.

Comment (TemporalOroboros) [reply Content.Client/Singularity/SingularityOverlay.cs:124]: Are you referring to the entire function or the final assignment of the screen position?

If you're referring to the former I have legitimately no idea how to convert the function into a matrix given that the amount of its scaling is dependent on the distance from the center of the singularity.

If you're referring to the latter I can do it pretty easily.

**Comment by AJCM-git** [comment]
reopen once you got time

**Comment by TemporalOroboros** [comment]
@AJCM-git 
Can you reopen this for me? I cant reopen it myself due to spam prevention. I've taken another look at it and corrected the math to actually work, but I'm still not certain what moon wants with the matrix math.

**Comment by TemporalOroboros** [comment]
Thanks.

**Comment by UbaserB** [comment]
This is actually more useful now, considering i am making a station which has active distortion effects even without singuloose, and i had several complaints including from atmos. I really hope this gets finished 👍🏻

**Comment by TemporalOroboros** [comment]
It could actually be shipped as is. It works, I'm just waiting for moony to come back and clarify what changes they want.

**Comment by ElectroJr** [comment]
I'm not sure what moony meant about the matrix stuff, but the distortion correction should probably not be part of the base `ScreenToMap()` function. Its probably better to make it some separate correction function that callers can use.

There are two main issues:
- It'd make the results of `ScreenToMap()` inconsistent with what is obtained by using the `GetWorldToScreenMatrix()` and `GetWorldToLocalMatrix()` matrix transformations.
- both `GetWorldViewbounds()` and `GetWorldViewport()` use `ScreenToMap()` and assume, which will probably cause issues when rendering near a singularity, especially when it heavily distorts one of the corners of the screen.


If its not made a separate function, the get-matrix functions need to be removed and the get-bounds functions needs to either not be distortion or the bounds calculation needs to be updated.

**Comment by TemporalOroboros** [comment]
@ElectroJr 
I've made separate functions for getting the origin of a visible pixel. They're in the engine PR. Does this work?

**Pull request merged**